src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs

```diff
@@ -126,6 +126,8 @@ public TagHelperBinder(string? tagHelperPrefix, ImmutableArray<TagHelperDescript
                 {
                     applicableDescriptors[descriptor] = applicableRules.DrainToImmutable();
                 }
+
+                applicableRules.Clear();
```

Comment:
I actually don't think this is necessary: if there were applicable rules, they would have been cleared by `DrainToImmutable`. That being said, I'm fine with keeping this for code clarity at the expense of a nanosecond or two.

Reply:
Oh right - I'd completely forgotten that `DrainToImmutable()` would effectively do that. It is pretty negligible though, so I'll leave it.